### PR TITLE
Issue 26972: Proper treatment of float values in intersectConvexConvex

### PR DESCRIPTION
**Pull request by klosteraner**
As outlined in https://github.com/opencv/opencv/issues/26972 the function `intersectConvexConvex()` may not work as expected in the corner case, where two polygons intersect at a corner. A concrete example is given that I added as unit test. The unit test would fail without the proposed bug fix. I recommend porting the fix to all versions.

Now concerning the fix: When digging into the implementation I found, that when the line intersections are computed, openCV currently does not apply floating point comparison syntax, but pretends that line end points are exact. Instead I replaced the formulation using the eps that is already used in another component of the function in line.277: `epx=1e-5`. IMO that is solid enough, definitely better than assuming an exact floating point comparison is possible.

As a follow up I would suggest to use a scalable eps, s.t. also cases with high floating point numbers would be less error prone. However that would need to be done in all relevant sub steps, not just the line intersection code. So for me outside the scope of this fix.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/imgproc/src/geometry.cpp [hunk 1]
```diff
@@ -346,8 +346,10 @@ static LineSegmentIntersection intersectLineSegments( Point2f a, Point2f b, Poin
     p.y = (float)(a.y + s*(b.y - a.y));
     q = p;
 
-    return s < 0. || s > 1. || t < 0. || t > 1. ? LS_NO_INTERSECTION :
-           s == 0. || s == 1. || t == 0. || t == 1. ? LS_ENDPOINT_INTERSECTION : LS_SINGLE_INTERSECTION;
+    static const double eps = 1e-5;
+    return s < - eps || s > 1.+ eps || t < - eps || t > 1. + eps ? LS_NO_INTERSECTION :
+           s < eps || s > 1. - eps || t < eps || t > 1. - eps ? LS_ENDPOINT_INTERSECTION :
+           LS_SINGLE_INTERSECTION;
```
Comment (asmorkalov) [inline modules/imgproc/src/geometry.cpp:352]: I agree that it's not correct to compare floats/doubles on equality, but why do you change the first line of conditions? I would say that the correct check is:
```
   return s < 0. || s > 1. || t < 0. || t > 1. ? LS_NO_INTERSECTION :
           fabs(num) < eps || fabs(s-1.) < eps || fabs(t) < eps || fabs(t-1.) < eps ? LS_ENDPOINT_INTERSECTION :
           LS_SINGLE_INTERSECTION;
```

Comment (asmorkalov) [reply modules/imgproc/src/geometry.cpp:352]: Also eps should be something like FLT_EPS or DBL_EPS.

Comment (klosteraner) [reply modules/imgproc/src/geometry.cpp:352]: The numerical error can happen on both sides, and in particular the one "outside" the line actually caused the problem that the intersection was not detected (see unit test). The one on the inside is probably harmless and could be treated either way, as endpoint and single intersection are "intersections". For symmetry reasons I also updated the second check.

**Comment by klosteraner** [comment]
The tests probably timed out due to unrelated reasons, can they be retriggered?

**Comment by klosteraner** [comment]
Anything missing from my side before merging?

**Comment by klosteraner** [comment]
So the original idea with putting the "numerical circle" around endpoints failed. I had examples, where a intersection was detected, even though polygons did not touch. For some reason the algorithm went crazy from that point on.

Now the new approach would be to only check "on the polygon". So if one edge goes through the vertex it will have its (let's say) s in [0,1]. Then the issue observed was that for the edge before the vertex we had t > 1 and after the vertex t < 0, but in fact we passed the vertex, so it should have been either t <= 1 or s>= 0. 

For now I put the MR to draft status: The original solution would not work, the new approach should circumvent those issues. I will remove the draft status once I arrive at a solution that reliably works for me. In the meantime I invite to iterate on the idea proposed with my second commit.



If you want to try yourself: 
    {-0.033171528879342396, -0.007706989388492961},
    {1920.2008671642784, -0.12243585190040425},
    {1920.2549093866564, 1080.028222999363},
    {0.020563744444937603, 1080.1050984957212}}
    and
   {{0, 0}, {1920, 0}, {1920, 1080}, {0, 1080}};

**Comment by klosteraner** [comment]
Ok, so unit tests pass now as before and also on my side I ran some initial automated tests with random rectangles - so far it looks good, so I would say the MR is again ready for review.

Compared to the first solution where I enlarged the vertex, I am now checking whether the intersecting lines start point was outside and its end point inside the polygon. Unless the edge is super tiny in the magnitude of the numeric errors, this is a robust alternative to checking whether a float point ==0.

I am not super happy yet with the names of variables and organization of the updates per "advance". I am open for suggestions making the code more readable. In fact if there is interest I could also propose some cleanup towards more readable code. It took me quite some time to wrap my head around the several a's b's p's and q's.

File: modules/imgproc/src/geometry.cpp [hunk 2]
```diff
@@ -346,8 +349,32 @@ static LineSegmentIntersection intersectLineSegments( Point2f a, Point2f b, Poin
     p.y = (float)(a.y + s*(b.y - a.y));
     q = p;
 
+    // Double check whether we had an intersection at the vertex, the == 0 check can fail
+    // It should be enough to check retrospectively - verify
+    const float eps = 1e-4;
```
Comment (klosteraner) [inline modules/imgproc/src/geometry.cpp:354]: With 1e-5 the unit test was still failing, so I went up to 1e-4. To keep it as small as possible, not more than that.

**Comment by klosteraner** [comment]
Can the tests be retriggered? Seems they did not run after removing the "Draft" status.

**Comment by vpisarev** [comment]
@klosteraner, thank you for the contribution and sorry for a long delay with the review. I found that if `intersectLineSegments()` is modified to use double-precision arithmetics more extensively, as shown below, then your test passes as-is, without complex modifications to intersectConvexConvex. If possible, could you please check it?

```
static LineSegmentIntersection intersectLineSegments( Point2f a, Point2f b, Point2f c,
                                                      Point2f d, Point2f& p, Point2f& q )
{
    double denom = ((double)a.x - b.x) * ((double)d.y - c.y) - ((double)a.y - b.y) * ((double)d.x - c.x);

    // If denom is zero, then segments are parallel: handle separately.
    if( denom == 0. )
        return parallelInt(a, b, c, d, p, q);

    double num = ((double)d.y - a.y) * ((double)a.x - c.x) + ((double)a.x - d.x) * ((double)a.y - c.y);
    double s = num / denom;

    num = ((double)b.y - a.y) * ((double)a.x - c.x) + ((double)c.y - a.y) * ((double)b.x - a.x);
    double t = num / denom;

    p.x = (float)(a.x + s*((double)b.x - a.x));
    p.y = (float)(a.y + s*((double)b.y - a.y));
    q = p;

    return s < 0. || s > 1. || t < 0. || t > 1. ? LS_NO_INTERSECTION :
           s == 0. || s == 1. || t == 0. || t == 1. ? LS_ENDPOINT_INTERSECTION : LS_SINGLE_INTERSECTION;
}
```

**Review by vpisarev — APPROVED** [review]

**Comment by klosteraner** [comment]
I see the reaction only now, thanks for the review. It is a fair point to keep the fix as minimal as possible, however I think the new fix is not a fix, but just better covering the underlying numerical problem by "making the holes smaller". It is not surprising that increasing the precision made the provided unit test pass. The issue however will remain the same just appear even less frequently.

It took >100k random examples to find an example that reproduces the issue before the fix, now its less frequent - probably good enough for the community. For my demands, I can live with continuing using the patch on my side only. To keep dependency handling more simple, I would appreciate though if you reconsider the original solution that changes the computation approach to have a robust computation that avoids that kind of numerical error completely.
